common/proxy_connection_handler.rb

```diff
@@ -18,22 +18,27 @@ def initialize(address, port)
       @address = address
       @port = port
       @reconnecting_socket = nil
+      @lock = Mutex.new
     end
 
     # Open a socket connection to the given address:port
     def connect
-      @reconnecting_socket = TCPSocket.open(@address, @port)
+      @lock.synchronize do
+        @reconnecting_socket = TCPSocket.open(@address, @port)
```

Comment:
one minor change - what if two threads get into synchronize block one after the other after checking initially that
@reconnecting_socket is nil.

so do this instead

@reconnecting_socket = TCPSocket.open(@address, @port) unless @reconnecting_socket.nil?